README.md
# NEP 29 Calculator


[![PyPI](https://img.shields.io/pypi/v/nep29.svg)](https://pypi.python.org/pypi/nep29)
[![Travis](https://img.shields.io/travis/hmaarrfk/nep29.svg)](https://travis-ci.com/hmaarrfk/nep29)


[NEP 29](https://numpy.org/neps/nep-0029-deprecation_policy.html) calculator tools

## Usage
```console
$ nep29 --help
usage: nep29 [-h] [--n_months N_MONTHS] [--n_minor N_MINOR] package

NEP 29 calculator.

positional arguments:
  package              Package to deprecation

optional arguments:
  -h, --help           show this help message and exit
  --n_months N_MONTHS  Number of months to keep supporting (default: 24)
  --n_minor N_MINOR    Number of minor versions to keep supporting (default: 3)

For more information, see https://numpy.org/neps/nep-0029-deprecation_policy.html
```

Example:

```console
$ nep29 numpy
| version |    date    |
|---------|------------|
|  1.19.0 | 2020-06-20 |
|  1.18.0 | 2019-12-22 |
|  1.17.0 | 2019-07-26 |
|  1.16.0 | 2019-01-14 |
```

## Credits

This package was created with [Cookiecutter](https://github.com/audreyr/cookiecutter)
and the
[hmaarrfk/cookiecutter-pypackage](https://github.com/hmaarrfk/cookiecutter-pypackage)
project template.

### Development Lead

* Mark Harfouche <mark.harfouche@gmail.com>

### Contributors

* Hugo van Kemenade (hugovk)

## Contributing

Contributions are welcome, and they are greatly appreciated! Every little bit
helps, and credit will always be given.

You can contribute in many ways:

### Report Bugs

Report bugs at https://github.com/hmaarrfk/nep29/issues.

If you are reporting a bug, please include:

* Your operating system name and version.
* Any details about your local setup that might be helpful in troubleshooting.
* Detailed steps to reproduce the bug.

### Fix Bugs

Look through the GitHub issues for bugs. Anything tagged with "bug" and "help
wanted" is open to whoever wants to implement it.

### Implement Features

Look through the GitHub issues for features. Anything tagged with "enhancement"
and "help wanted" is open to whoever wants to implement it.

### Write Documentation

NEP29 Calculator could always use more documentation, whether as part of the
official NEP29 Calculator docs, in docstrings, or even on the web in blog posts,
articles, and such.

### Submit Feedback

The best way to send feedback is to file an issue at https://github.com/hmaarrfk/nep29/issues.

If you are proposing a feature:

* Explain in detail how it would work.
* Keep the scope as narrow as possible, to make it easier to implement.
* Remember that this is a volunteer-driven project, and that contributions
  are welcome :)

### Get Started!

You can follow the scikit-image contribution document and adapt it accordingly.



HISTORY.md
# History

## 0.0.17 (2022-11-29)

* Skip invalid versions instead of erroring.

## 0.0.16 (2022-01-30)

* Add tests back to distribution.

## 0.0.15 (2022-01-30)

* Use miniver instead of versioneer.

## 0.0.10 (2021-12-30)

* Add versioneer.py to the source distribution.

## 0.0.9 (2021-11-17)

* Retrigger failed build due to travis.org migration.

## 0.0.8 (2021-06-16)

* Add support for pre-release packages that don't use a dot to separate the pre-release signifier.

## 0.0.7 (2021-11-03)

* Show default argument values in the help.

## 0.0.6 (2021-11-03)

* Support a few more time formats

## 0.0.5 (2020-10-13)

* Use pretty table for better formatting.

## 0.0.4 (2020-10-12)

* Improve documentation.

## 0.0.1 (2020-03-11)

* First release on PyPI.


nep29/tests/__init__.py
"""Unit test package for nep29."""


nep29/tests/test_nep29.py
import nep29
from nep29 import nep29_versions
import pytest
from unittest.mock import patch, Mock
from datetime import datetime
from packaging.version import Version


def test_project_import():
    pass


@pytest.mark.parametrize('vstring,expected', [
    ("1!2.3.4", (1, 2, 3)),
    ("1.2.3", (0, 1, 2)),
])
def test_major_minor_version(vstring, expected):
    from packaging.version import Version
    v = Version(vstring)
    assert nep29.nep29.major_minor_version(v) == expected


def _mock_release_info(release_info):
    # returns a mock for the requests.Response object
    def _make_release_entry(version, date, packagetype='sdist'):
        return version, [{
            'packagetype': packagetype,
            'upload_time_iso_8601': date + 'T00:00:00.000000Z'
        }]

    package = {'releases': dict(_make_release_entry(*info)
                                for info in release_info)}
    request = Mock()
    request.json = Mock(return_value=package)
    return request


@patch.object(nep29.nep29, 'datetime', Mock(wraps=datetime))
def _nep29_versions_test(request_mock, expected, *args, **kwargs):
    # core code to run tests of nep29_versions from a mock for the
    # requests.Response object as a list of expected version strings;
    # additional args/kwargs passed to nep29_versions
    nep29.nep29.datetime.utcnow.return_value = datetime(2021, 1, 1)
    with patch('nep29.nep29.requests.get', Mock(return_value=request_mock)):
        results = nep29_versions('foo', *args, **kwargs)

    found_versions = set(r[0] for r in results)
    assert found_versions == set(Version(v) for v in expected)


@pytest.mark.parametrize('releases,expected', [
    # NOTE: for these tests we patch so that today is 2021-01-01, midnight.
    # All mock releases are at midnight GMT.
    (
        # default: include up to 3 minor versions if released long ago
        _mock_release_info([('2.0.0', '2015-01-01'),
                            ('1.2.0', '2014-01-01'),
                            ('1.1.0', '2013-01-01'),
                            ('1.0.0', '2012-01-01')]),
        ['2.0.0', '1.2.0', '1.1.0']
    ),
    (
        # default: include all minor versions first released in 24 months
        _mock_release_info([('1.9.0', '2020-12-01'),
                            ('1.8.0', '2020-09-01'),
                            ('1.7.0', '2020-06-01'),
                            ('1.6.0', '2020-01-01'),
                            ('1.5.0', '2019-06-01'),
                            ('1.4.0', '2018-12-01')]),
        ['1.9.0', '1.8.0', '1.7.0', '1.6.0', '1.5.0']
    ),
    (
        # default: release candidates should not be included
        _mock_release_info([('2.0.0.rc0', '2020-12-01'),
                            ('1.3.0', '2020-11-01'),
                            ('1.2.0', '2020-10-01'),
                            ('1.1.0', '2020-09-01')]),
        ['1.3.0', '1.2.0', '1.1.0']
    ),
])
def test_nep29_versions(releases, expected):
    _nep29_versions_test(releases, expected)


def test_nep29_versions_allow_rc():
    releases = _mock_release_info([('2.0.0.rc0', '2020-12-01'),
                                   ('1.3.0', '2020-11-01'),
                                   ('1.2.0', '2020-10-01'),
                                   ('1.1.0', '2020-09-01')])
    expected = ['2.0.0.rc0', '1.3.0', '1.2.0', '1.1.0']
    _nep29_versions_test(releases, expected, skip_rc=False)


@pytest.mark.parametrize('package', [
    # Pre-release information can be included as letters
    # not seperated by a '.' from the version number
    # the pandas package uses this notation.
    'pandas',
    # h5py contains a pre-release
    # With something like an invalid version
    #       Invalid version: '1.3.0.dev-r634'
    'h5py',
])
def test_pypi_package(package):
    nep29_versions(package)


nep29/_version.py
# -*- coding: utf-8 -*-
# This file is part of 'miniver': https://github.com/jbweston/miniver
#
from collections import namedtuple
import os
import subprocess

from setuptools.command.build_py import build_py as build_py_orig
from setuptools.command.sdist import sdist as sdist_orig

Version = namedtuple("Version", ("release", "dev", "labels"))

# No public API
__all__ = []

package_root = os.path.dirname(os.path.realpath(__file__))
package_name = os.path.basename(package_root)

STATIC_VERSION_FILE = "_static_version.py"


def get_version(version_file=STATIC_VERSION_FILE):
    version_info = get_static_version_info(version_file)
    version = version_info["version"]
    if version == "__use_git__":
        version = get_version_from_git()
        if not version:
            version = get_version_from_git_archive(version_info)
        if not version:
            version = Version("unknown", None, None)
        return pep440_format(version)
    else:
        return version


def get_static_version_info(version_file=STATIC_VERSION_FILE):
    version_info = {}
    with open(os.path.join(package_root, version_file), "rb") as f:
        exec(f.read(), {}, version_info)
    return version_info


def version_is_from_git(version_file=STATIC_VERSION_FILE):
    return get_static_version_info(version_file)["version"] == "__use_git__"


def pep440_format(version_info):
    release, dev, labels = version_info

    version_parts = [release]
    if dev:
        if release.endswith("-dev") or release.endswith(".dev"):
            version_parts.append(dev)
        else:  # prefer PEP440 over strict adhesion to semver
            version_parts.append(".dev{}".format(dev))

    if labels:
        version_parts.append("+")
        version_parts.append(".".join(labels))

    return "".join(version_parts)


def get_version_from_git():
    # git describe --first-parent does not take into account tags from branches
    # that were merged-in. The '--long' flag gets us the 'dev' version and
    # git hash, '--always' returns the git hash even if there are no tags.
    for opts in [["--first-parent"], []]:
        try:
            p = subprocess.Popen(
                ["git", "describe", "--tags", "--long", "--always"] + opts,
                cwd=package_root,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
            )
        except OSError:
            return
        if p.wait() == 0:
            break
    else:
        return

    description = (
        p.communicate()[0]
        .decode()
        .strip("v")  # Tags can have a leading 'v', but the version should not
        .rstrip("\n")
        .rsplit("-", 2)  # Split the latest tag, commits since tag, and hash
    )

    try:
        release, dev, git = description
    except ValueError:  # No tags, only the git hash
        # prepend 'g' to match with format returned by 'git describe'
        git = "g{}".format(*description)
        release = "unknown"
        dev = None

    labels = []
    if dev == "0":
        dev = None
    else:
        labels.append(git)

    try:
        p = subprocess.Popen(["git", "diff", "--quiet"], cwd=package_root)
    except OSError:
        labels.append("confused")  # This should never happen.
    else:
        if p.wait() == 1:
            labels.append("dirty")

    return Version(release, dev, labels)


# TODO: change this logic when there is a git pretty-format
#       that gives the same output as 'git describe'.
#       Currently we can only tell the tag the current commit is
#       pointing to, or its hash (with no version info)
#       if it is not tagged.
def get_version_from_git_archive(version_info):
    try:
        refnames = version_info["refnames"]
        git_hash = version_info["git_hash"]
    except KeyError:
        # These fields are not present if we are running from an sdist.
        # Execution should never reach here, though
        return None

    if git_hash.startswith("$Format") or refnames.startswith("$Format"):
        # variables not expanded during 'git archive'
        return None

    VTAG = "tag: v"
    refs = set(r.strip() for r in refnames.split(","))
    version_tags = set(r[len(VTAG):] for r in refs if r.startswith(VTAG))
    if version_tags:
        release, *_ = sorted(version_tags)  # prefer e.g. "2.0" over "2.0rc1"
        return Version(release, dev=None, labels=None)
    else:
        return Version("unknown", dev=None, labels=["g{}".format(git_hash)])


__version__ = get_version()


# The following section defines a 'get_cmdclass' function
# that can be used from setup.py. The '__version__' module
# global is used (but not modified).


def _write_version(fname):
    # This could be a hard link, so try to delete it first.  Is there any way
    # to do this atomically together with opening?
    try:
        os.remove(fname)
    except OSError:
        pass
    with open(fname, "w") as f:
        f.write(
            "# This file has been created by setup.py.\n"
            "version = '{}'\n".format(__version__)
        )


def get_cmdclass(pkg_source_path):
    class _build_py(build_py_orig):
        def run(self):
            super().run()

            src_marker = "".join(["src", os.path.sep])

            if pkg_source_path.startswith(src_marker):
                path = pkg_source_path[len(src_marker):]
            else:
                path = pkg_source_path
            _write_version(
                os.path.join(
                    self.build_lib, path, STATIC_VERSION_FILE
                )
            )

    class _sdist(sdist_orig):
        def make_release_tree(self, base_dir, files):
            super().make_release_tree(base_dir, files)
            _write_version(
                os.path.join(base_dir, pkg_source_path, STATIC_VERSION_FILE)
            )

    return dict(sdist=_sdist, build_py=_build_py)


nep29/nep29.py
"""
# NEP 29

NEP 29 reads

> When a project releases a new major or minor version, we recommend that they
> support at least [...] all minor versions of NumPy released in the prior
> 24 months from the anticipated release date with a minimum of 3 minor
> versions of NumPy

This means we have to create the union of two sets:

1. Releases that are in the last X months -- only consider X.X.0 releases
2. A certain number of minor versions.

Copyright Mark Harfouche 2019
"""
from datetime import datetime, timedelta
from prettytable import PrettyTable, MARKDOWN
import requests
from packaging.version import Version, InvalidVersion


def major_minor_version(version):
    """Given a version, returns the tuple epoch, major, minor"""
    # version.release works in either packaging 19 or 20
    # as of packaging 20+, can use version.major, version.minor instead
    major, minor = version.release[:2]
    return version.epoch, major, minor


def keep_oldest_minor_only(version_dates):
    vd_dict = {}
    for version, date in version_dates:
        major_minor = major_minor_version(version)
        if major_minor in vd_dict:
            if vd_dict[major_minor][1] > date:
                vd_dict[major_minor] = (version, date)
        else:
            vd_dict[major_minor] = (version, date)
    return [val
            for _, val in vd_dict.items()]


def get_versions_dates(package_name, skip_rc=True):
    r = requests.get(f"https://pypi.org/pypi/{package_name}/json")
    response = r.json()
    release_dates = []
    for k, v in response['releases'].items():
        # NOTE: for a version that doesn't comply to PEP 440, this will
        # raise packaging.version.InvalidVersion:
        try:
            version = Version(k)
        except InvalidVersion:
            pass
        for item in v:
            if item['packagetype'] == 'sdist':
                if version.is_prerelease:
                    if skip_rc or not version.pre[0] == 'rc':
                        # skip unless we do not skip RCs and this is RC
                        continue
                # alt: interpret skip_rc as allow any prerelease (e.g, betas)
                # if skip_rc and version.is_prerelease:
                    # continue
                upload_time = item['upload_time_iso_8601']
                for format in ['%Y-%m-%dT%H:%M:%S.%fZ',
                               '%Y-%m-%dT%H:%M:%SZ']:
                    try:
                        date = datetime.strptime(upload_time, format)
                        break
                    except ValueError:
                        pass
                else:
                    raise ValueError(
                        f"Could not convert {upload_time} into a standard "
                        "datetime object")

                release_dates.append((version, date))

    release_dates.sort(key=lambda x: x[1], reverse=True)
    return release_dates


def good_nep29_date(version_dates, n_months=24, release_date=None):
    delta_time = timedelta(days=365 / 12 * n_months)

    if release_date is None:
        release_date = datetime.utcnow()

    oldest_release_date = release_date - delta_time
    good_versions_date = [vd[1] > oldest_release_date
                          for vd in version_dates]
    return good_versions_date


def good_nep29_minor(version_dates, n_minor=3):
    major_minor_version_dates = [(major_minor_version(v), date)
                                 for v, date in version_dates]

    unique_major_minor = set()
    n_unique_major_minor = []
    for mm, date in major_minor_version_dates:
        unique_major_minor.add(mm)
        n_unique_major_minor.append(len(unique_major_minor))

    # Find out if it is a new "minor" version
    new_major_minor = [bool(n_current - n_old)
                       for n_current, n_old in zip(n_unique_major_minor[1:],
                                                   n_unique_major_minor[:-1])]
    # The very newest release is always "new"
    new_major_minor = [1] + new_major_minor

    good_versions_minor = [n <= n_minor and new
                           for n, new in zip(n_unique_major_minor,
                                             new_major_minor)]
    return good_versions_minor


def nep29_versions(package_name, *,
                   n_months=24, n_minor=3,
                   skip_rc=True,
                   release_date=None,
                   consider_first_minor_only=True):
    version_dates = get_versions_dates(package_name, skip_rc=skip_rc)
    if consider_first_minor_only:
        version_dates = keep_oldest_minor_only(version_dates)
    good_nep29_date_indicator = good_nep29_date(version_dates,
                                                n_months=n_months,
                                                release_date=release_date)
    good_nep29_minor_indicator = good_nep29_minor(version_dates,
                                                  n_minor=n_minor)

    # Take the combination of the two sets
    good_indicator = [g1 or g2
                      for g1, g2 in zip(good_nep29_date_indicator,
                                        good_nep29_minor_indicator)]

    valid_releases = [vd
                      for vd, good in zip(version_dates, good_indicator)
                      if good]

    valid_releases.sort(reverse=True)
    return valid_releases


def main():
    import argparse
    parser = argparse.ArgumentParser(
            description="NEP 29 calculator.",
            epilog=(
                "For more information, see "
                "https://numpy.org/neps/nep-0029-deprecation_policy.html"
            ),
            formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('package', type=str, help='Package to deprecation')
    parser.add_argument('--n_months', type=int, default=24,
                        help='Number of months to keep supporting')
    parser.add_argument('--n_minor', type=int, default=3,
                        help='Number of minor versions to keep supporting')
    args = parser.parse_args()

    package = args.package
    n_months = args.n_months
    n_minor = args.n_minor
    version_dates = nep29_versions(package, n_months=n_months, n_minor=n_minor)
    t = PrettyTable(["version", "date"])
    t.set_style(MARKDOWN)
    for version, date in version_dates:
        t.add_row([version, date.strftime('%Y-%m-%d')])
    print(t)


"""
from pprint import pprint
print("SciPy NEP 29 requirements")
pprint(nep29_versions('scipy', n_months=24, n_minor=3))

print("NumPy NEP 29 requirements")
pprint(nep29_versions('numpy', n_months=24, n_minor=3))
"""
